### PR TITLE
Add `accessPropertiesByDotNotation` option

### LINKED ISSUE #44: Add feature to disable dot-notation
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/conf/issues/44)

Hi! I'm working on a project that uses reverse-domain naming for namespaces, and we need to store some data with the namespace as the key. Where we'd expect 
```JSON
{
    "cc.xylobol.test": {
        "foo": "bar"
    }
}
```
we end up with 
```JSON
{
    "cc": {
        "xylobol": {
            "test": {
                "foo": "bar"
            }
        }
    }
}
```
instead due to the dot-notation. Is there a pre-existing way to disable this? If not, could it be added as a feature?

Thank you in advanced.
Xylobol

> There is a $40.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/conf/issues/44).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/conf/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by yaodingyd**
fix #44

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -188,6 +188,13 @@ You can pass an empty string to remove the suffix.
 
 For example, on macOS, the config file will be stored in the `~/Library/Preferences/foo-nodejs` directory, where `foo` is the `projectName`.
 
+#### disableDotNotation
```
Comment (sindresorhus) [inline readme.md]: Double-negative option names are not optimal. Would be better with `dotNotation`.

File: readme.md [hunk 2]
```diff
@@ -188,6 +188,13 @@ You can pass an empty string to remove the suffix.
...
+
+Type: `boolean`<br>
+Default: `false`
+
+Disable dot-notation for accessing nested properties, treat the whole string as one key.
```
Comment (sindresorhus) [inline readme.md]: Add an example. Might not be clear to users what `dot-notation` means.

File: index.js [hunk 3]
```diff
@@ -110,8 +110,14 @@ class Conf {
...
 
+	_getkey(key) {
+		const escapeDot = value => value.replace(/\./g, '\\\\.');
+
+		return this._options.disableDotNotation ? escapeDot(key) : key;
+	}
+
 	get(key, defaultValue) {
-		return dotProp.get(this.store, key, defaultValue);
+		return dotProp.get(this.store, this._getkey(key), defaultValue);
```
Comment (sindresorhus) [inline index.js]: Instead of escaping the dot, simply don't use `dotProp` when dot-notation is turned off.

**Comment by sindresorhus** [comment]
I'm not super happy about the name `dotNotation`. Can you list some other name ideas for the option?

**Comment by yaodingyd** [comment]
Most explicit would be `accessPropertyByDotDonation`

File: index.d.ts [hunk 4]
```diff
@@ -125,6 +125,45 @@ declare namespace Conf {
...
+
+		config.set({
+			`foo.bar.foobar`: '🦄'
+		});
+
+		console.log(config.get('foo.bar.foobar'));
+		//=> '🦄'
+		```
+
+		@default 'true'
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
		@default true
```

File: index.d.ts [hunk 5]
```diff
@@ -125,6 +125,45 @@ declare namespace Conf {
 		@default 'nodejs'
 		*/
 		readonly projectSuffix?: string;
+
+		/**
+		Accessing nested properties by dot nonation. For example:
```
Comment (sindresorhus) [inline index.d.ts]: Typo

File: test.js [hunk 6]
```diff
@@ -11,6 +11,7 @@ const fixture = '🦄';
 
 test.beforeEach(t => {
 	t.context.conf = new Conf({cwd: tempy.directory()});
+	t.context.config = new Conf({cwd: tempy.directory(), accessPropertiesByDotNotation: false});
```
Comment (sindresorhus) [inline test.js]: It's weird that this is called `config` and the other one `conf`. Be more specific.

File: index.d.ts [hunk 7]
```diff
@@ -125,6 +125,45 @@ declare namespace Conf {
...
+
+		console.log(config.get('foo.bar.foobar'));
+		//=> '🦄'
+		```
+
+		@default true
+		*/
+
+		readonly accessPropertiesByDotNotation?: boolean;
+
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
```

File: index.d.ts [hunk 8]
```diff
@@ -125,6 +125,45 @@ declare namespace Conf {
...
+			`foo.bar.foobar`: '🦄'
+		});
+
+		console.log(config.get('foo.bar.foobar'));
+		//=> '🦄'
+		```
+
+		@default true
+		*/
+
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
```

File: index.d.ts [hunk 9]
```diff
@@ -125,6 +125,45 @@ declare namespace Conf {
...
+		config.set({
+			`foo.bar.foobar`: '🦄'
+		});
+
+		console.log(config.get('foo.bar.foobar'));
+		//=> '🦄'
+		```
+
+		@default true
```
Comment (sindresorhus) [inline index.d.ts]: This should be before the example.

File: index.d.ts [hunk 10]
```diff
@@ -125,6 +125,45 @@ declare namespace Conf {
 		@default 'nodejs'
 		*/
 		readonly projectSuffix?: string;
+
+		/**
+		Accessing nested properties by dot notation. For example:
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
		Access nested properties by dot notation.
```

**Comment by sindresorhus** [comment]
I feel like the option is a bit too verbose, but I can't really think of anything else, and at least it's clearer than `dotNotation`.

**Comment by sindresorhus** [comment]
A new release is out. Can you document this PR and the other one in `electron-store` too? https://github.com/sindresorhus/conf/releases/tag/v4.1.0

**Comment by yaodingyd** [comment]
sure, on it

**Pull request merged**